Add syntax to the parser for local allocations

lpw25 opened the pull request:
Add some syntax for local allocations.

This PR adds a new expression form:
```ocaml
stack_  expr
```
which forces the mode of the result to be local.

It also adds new forms for parameters:
```ocaml
fun (local_ pat) ~(local_ x) ~x:(local_ pat) ?(local_ x) ?x:(local_ pat) -> ...
```
that make the parameters local.

It also adds a variant of `stack_` for functions defined with `let`:
```ocaml
let stack_ foo x = ...
```
that makes the function closure local.

It adds syntax for local parameters in types:
```ocaml
local_ string -> local_ float -> int
```
and for the types of functions that return local values:
```ocaml
int -> int -> stack_ string
```

The PR also makes the parser treat `t -> t -> t` differently from `t -> (t -> t)`. That allows us to have `local_ 'a -> 'b -> 'c` be treated as `local_ 'a -> stack_ ('b -> 'c)`, whilst `local_ 'a -> ('b -> 'c)` leaves the `'b -> 'c` part at global mode.

Finally, the PR also makes the parser treat `fun x y -> expr` differently from `fun x -> fun y -> expr`. This allows is to have `fun (local_ x) y -> ...` produce a value of type `local 'a -> 'b -> 'c`, whilst `fun x -> fun y -> ...` will try to produce a value of type `local 'a -> ('b -> 'c)`.

lpw25 committed:
Add syntax to the parser for local allocations

lpw25 committed:
Lower initial stack size

lpw25 commented:
I've rebased the PR on top of the mode variables patch, and I've updated it so that it now also prevents arguments to tail calls from being local, and only allows function returns to be local if they are annotated with `stack_`.

stedolan commented on testsuite/tests/parsing/multi_indices.ml:
```diff
@@ -35,8 +35,7 @@ val ( .![;..]<- ) : ('a, 'b, 'c) A.t -> int array -> 'a -> unit = <fun>
   Format.printf "indices: @[[|%a|]@]@."
-    (Format.pp_print_list
-       ~pp_sep:(fun ppf -> fun () -> Format.fprintf ppf ";@ ")
+    (Format.pp_print_list ~pp_sep:(fun ppf  () -> Format.fprintf ppf ";@ ")
```
Was this change needed for compatibility? What goes wrong with the original?

lpw25 replied:
Now that we are tracking the difference between `fun x y ->` and `fun x -> fun y ->` the pretty printer needs to respect that difference. Note that the new output matches the input.

stedolan replied:
Makes sense! I hadn't spotted that this was in tests/parsing, I thought there might be some typing issue here.

stedolan commented on typing/types.ml:
```diff
@@ -593,6 +598,21 @@ module Alloc_mode = struct
     vars |> List.iter (fun v' -> submode_vv v' v);
     v
 
+  let join2 m1 m2 =
```
There's a lot of code duplication here. Either this is `join [m1;m2]` or join is `fold_left join2`.

stedolan commented on runtime/memory.c:
```diff
@@ -692,7 +692,7 @@ CAMLexport void caml_local_region_end(intnat reg)
 
 //#define Local_init_wsz 64
-#define Local_init_wsz (64*1024*1024)
+#define Local_init_wsz (4096)
```
Is this relevant?

lpw25 replied:
No, but I needed it to stop `make parallel` on the testsuite from getting OOM killed.

stedolan commented on typing/typemod.ml:
```diff
@@ -1957,6 +1957,7 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
           in
           Named (id, param, mty), Types.Named (id, mty.mty_type), newenv, true
       in
+      let newenv = Env.add_lock Alloc_mode.heap newenv in
```
What's this for?

stedolan replied:
Thinking about it some more, this looks like an attempt to prevent local allocations leaking out through functors defined inside expressions that make local allocations.

This could be made to work, but keeping modes correct throughout the module language will be tricky and I don't think it's worthwhile. Instead, I've been trying to ensure that the environment is always fully locked (no accessible locals) when typing anything in the module language.

So I think there is indeed a missing lock, but not here: there should be a lock in Pexp_letmodule where we switch to the module language and nothing in typemod.

lpw25 replied:
I don't think we need to be that restrictive. The rules for the module language are easy:
1. structs are all on the heap, when you put values into a struct you need to check that the mode is heap
2. functors are closures on the heap, so the context needs to be locked for their bodies
3. Since all modules are on the heap, unpacking a first-class module must check that its mode is heap

Generally, you only need to lock when you have expressions whose execution is not immediate -- i.e. they are going to be in a closure. Otherwise we only need to ensure that the expected modes are correct.

stedolan commented:
My understanding is that `let stack_ f x = ...` is the same as `let f = stack_ fun x -> ...`. Is the `let stack_ ...` syntax supported for anything other than function bindings?

stedolan commented on typing/typecore.ml:
```diff
@@ -428,6 +456,7 @@ let enter_variable ?(is_module=false) ?(is_as_variable=false) loc name mode ty
     (* Note: unpack patterns enter a variable of the same name *)
     if not !allow_modules then
       raise (Error (loc, Env.empty, Modules_not_allowed));
+    escape ~loc ~env:Env.empty mode;
```
I don't understand this check.

lpw25 replied:
If you unpack a first-class module then it should be heap, since we do not track whether modules escape.

stedolan commented on typing/typedtree.mli:
```diff
@@ -49,6 +49,7 @@ and 'a pattern_data =
     pat_loc: Location.t;
     pat_extra : (pat_extra * Location.t * attributes) list;
     pat_type: Types.type_expr;
+    pat_mode: Types.alloc_mode;
```
What's this for?

lpw25 replied:
This mean you can tell what mode a variable is bound at by looking at the typedtree. We use that for checking what escapes in top-level bindings, but seems a useful property in general for the typedtree.

stedolan commented on typing/typecore.ml:
```diff
@@ -1959,15 +1998,7 @@ let type_pattern_list
-    let mode =
-      match stack_annot, no_existentials with
-      | true, At_toplevel ->
-         raise (Error(pat.ppat_loc, env, Local_value_escapes))
```
Where has this check gone? The point is to ensure that module-level bindings are not inferred as "local".

lpw25 replied:
Rather than insisting that all top-level bindings are heap, we check that things are heap at the point that they are added to a struct. It's not much harder to do the check that way and it lines up more closely with the semantics. It is also more expressive although not in an especially useful way:
```ocaml
let { contents = x } = stack_ ref 5
```

stedolan reviewed (changes requested):
This is looking good!

I haven't finished reading typecore yet, but it's probably better to show the first batch of comments sooner.

lpw25 commented:
> Is the let stack_ ... syntax supported for anything other than function bindings?

Only function bindings for now. I'm open to additional syntax, but was trying to keep things fairly minimal in this PR.

stedolan commented on testsuite/tests/typing-local/local.ml:
```diff
@@ -192,152 +170,154 @@ val apply4 : int -> int = <fun>
-[%%expect{|
-type ('a, 'b) b_sfn = b:('a [@stack]) -> 'b
-val bug2 : unit -> c:int -> unit = <fun>
+  let foo : a:local_ string -> (b:local_ string -> c:int -> unit) =
```
I think that this test has changed. It's supposed to be:
```
  let foo : a:local_ string -> (b:local_ string -> (c:int -> unit))
```
That is, once `~a` and `~b` are supplied, the result is a heap closure of type `c:int -> unit`, and can validly be returned. (The test checks that this works even if `~a` and `~b` are supplied backwards via partial application).

stedolan commented on testsuite/tests/typing-local/local.ml:
```diff
@@ -192,152 +170,154 @@ val apply4 : int -> int = <fun>
 |}]
 let bug3 () =
-  let foo : a:(string[@stack]) -> (string, c:int -> unit) b_sfn = fun ~a ~b -> fun ~c -> print_string a in
+  let foo : a:local_ string -> (b:local_ string -> c:int -> unit) =
```
Same issue as above

stedolan commented on typing/typecore.ml:
```diff
@@ -3350,6 +3412,7 @@ and type_expect_
           | _ ->
               assert false
         in
+        let _mode = unannotated_mode ~loc ~env mode in
```
This looks odd. I presume this is part of the FIXME in exp_mode below?

stedolan commented:
I see you've chosen to encode `let x = local_ e` as an extension `[%stack e]` rather than an attribute `(e [@stack])`. This choice will break PPXes that try to rewrite expressions (IIRC, they mostly don't descend into extension nodes), but maybe that's OK?

stedolan commented on typing/typecore.ml:
```diff
@@ -3842,14 +3939,14 @@ and type_function ?in_function loc attrs env mode ty_expected_explained l caseli
-    | Error _ -> raise (Error(loc, env, Local_return_value_escapes))
+  if has_local then
+    eqmode ~loc ~env arg_mode alloc_local (Param_mode_clash ty_expected');
+  begin
+    match uncurried_function with
+    | None -> ()
+    | Some inner_loc ->
+        submode ~loc:inner_loc ~env arg_mode ret_mode Local_value_escapes;
+        submode ~loc ~env mode ret_mode Local_value_escapes
   end;
```
I'm trying to wrap my head around the logic here but I don't get it.

stedolan replied:
(It was explained, I got it. Short version: in a curried function, the outer lambda constructs a partial application. The value returned by the outer lambda is therefore a closure, closing over the argument and the function itself, and so both the arg mode and ambient mode of the outer lambda must be below its return mode)

stedolan reviewed (approved):
LGTM.

This patch introduces a lot of new typing checks but almost no new tests. If you want to add these in a future PR, that's fine by me, but they do need to appear eventually. (Leaving them until after the upcoming GC changes makes sense: the testsuite will be in better shape then, and more able to detect silent escapes via heap allocations)

Here's a (partial?) list of new checks that should be hit by the testsuite:

  - there are new checks for escaping locals through lazy values
  - the new weaker module typing rules mean there are new checks to ensure things don't escape (via functors, classes, first-class modules...)
  - class typing has new subtle checking around parameters and instance variables
  - tail calls now check their arguments differently
  - unannotated local returns are now prevented unless they are uncurried function definitions
  - parameter modes are checked for clashes between type and term-level specifications

lpw25 committed:
Bootstrap

lpw25 commented:
I switched to an internal representation of `[%stack] expr` instead of `[%stack expr]`. That should allow ppxes to work whilst still ensuring that we don't forget to handle any instances.

I've also added tests for all the cases you mentioned.
